lib/ABI/FunctionTypeConversion.cpp

```diff
@@ -0,0 +1,387 @@
+      // is gathered).
+      Result.emplace_back().Type() = { defaultRegisterType(Register), {} };
+    }
+  }
+
+  // Rewrite all the indices to make sure they are consistent.
+  for (auto Pair : llvm::enumerate(Result))
+    Pair.value().Index() = Pair.index();
+
+  return Result;
+}
+
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                                       size_t IndexOffset) const {
+  // Qualifiers here are not allowed since this must point to a simple struct.
+  revng_assert(StackArgumentTypes.Qualifiers().empty());
+
+  auto *Unqualified = StackArgumentTypes.UnqualifiedType().get();
+  if (!Unqualified) {
+    // If there is no type, it means that the importer responsible for this
+    // function didn't detect any stack arguments and avoided creating
+    // a new empty type.
+    return llvm::SmallVector<model::Argument, 8>{};
+  }
+
+  // If the struct is empty, it indicates that there are no stack arguments.
+  const model::StructType &Types = *llvm::cast<model::StructType>(Unqualified);
+  if (Types.Fields().empty()) {
+    revng_assert(Types.size() == 0);
+    return llvm::SmallVector<model::Argument, 8>{};
+  }
+
+  // Compute the alignment of the first argument.
+  auto CurrentAlignment = *ABI.alignment(Types.Fields().begin()->Type());
+  if (!isZeroOrPowerOfTwo(CurrentAlignment)) {
+    revng_log(Log,
+              "The natural alignment of a type is not a power of two:\n"
+                << serializeToString(Types.Fields().begin()->Type()));
+    return std::nullopt;
+  }
+
+  // Process each of the fields (except for the very last one),
+  // converting them into arguments separately.
+  llvm::SmallVector<model::Argument, 8> Result;
+  for (auto Iterator = Types.Fields().begin();
+       Iterator != std::prev(Types.Fields().end());
+       ++Iterator) {
+    const auto &CurrentArgument = *Iterator;
+    auto MaybeSize = CurrentArgument.Type().size();
+    revng_assert(MaybeSize.has_value() && MaybeSize.value() != 0);
+
+    const auto RegisterSize = model::ABI::getPointerSize(ABI.ABI());
+    uint64_t CurrentSize = paddedSizeOnStack(MaybeSize.value(), RegisterSize);
+
+    const auto &NextArgument = *std::next(Iterator);
+    auto NextAlignment = *ABI.alignment(NextArgument.Type());
+    if (!isZeroOrPowerOfTwo(NextAlignment)) {
+      revng_log(Log,
+                "The natural alignment of a type is not a power of two:\n"
+                  << serializeToString(NextArgument.Type()));
+      return std::nullopt;
+    }
+
+    OverflowSafeInt Offset = CurrentArgument.Offset();
+    Offset += CurrentSize;
+    if (!Offset) {
+      // Abandon if offset overflows.
+      revng_log(Log, "Integer overflow when calculating field offsets.");
+      return std::nullopt;
+    }
+
+    if (*Offset != NextArgument.Offset() || *Offset % NextAlignment != 0) {
```

Comment:
Actually this code was broken and I had fix for a while, sorry about not pushing it fast enough so you had to look at the old version. Here it is: https://github.com/revng/revng/pull/232/commits/ecdcae6b135b0e6c4a4ac339f16efd51f7d7e83b

This solves the issues you brought up as `2` and `3` (which I also already ran into earlier)

As for 1, the whole point of this is to verify how `CurrentArgument` relates to `NextArgument`.

In simple terms, what happens here is that we get the position of the `CurrentArgument` and add its size (with applied padding as well taking the alignment of the `NextArgument` into the account), this way we try to predict where the `NextArgument` is supposed to be placed based on their types. And then it looks if it's really there, if not: abort. See the new version, it should be a lot easier to follow now.

Reply:
I'd gladly assume that this discussion is also over, but I feel like it's better if you take one more look at the newest version (https://github.com/revng/revng/pull/232/commits/730a1c54cff9f790f5156c18e5eb4f620ae50e1d#diff-451550659236a2555b9c4555b7f6d90e15845726ba9329c6922795a8078fe5b8R221-R234) just in case.

Does the check make sense now?